pkg/validation-webhook/snapshot.go

```diff
@@ -0,0 +1,203 @@
+		if _, _, err := deserializer.Decode(raw, nil, snapcontent); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		oldSnapcontent := &volumesnapshotv1beta1.VolumeSnapshotContent{}
+		if _, _, err := deserializer.Decode(oldRaw, nil, oldSnapcontent); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		return decideSnapshotContent(snapcontent, oldSnapcontent, isUpdate)
+	default:
+		err := fmt.Errorf("expect resource to be %s or %s", SnapshotGVR, SnapshotContentGVR)
+		klog.Error(err)
+		return toV1AdmissionResponse(err)
+	}
+}
+
+func decideSnapshot(snapshot, oldSnapshot *volumesnapshotv1beta1.VolumeSnapshot, isUpdate bool) *v1.AdmissionResponse {
+	reviewResponse := v1.AdmissionResponse{
+		Allowed: true,
+		Result:  &metav1.Status{},
+	}
+
+	// if it is an UPDATE and oldSnapshot is not valid, then don't enforce strict validation
+	// This allows no-op updates to occur on snapshot resources which fail strict validation
+	// Which allows the remover of finalizers and therefore deletion of this object
+	// Don't rely on the pointers to be nil, because the deserialization method will convert it to
+	// The empty struct value. Instead check the operation type.
+	err := utils.ValidateSnapshot(oldSnapshot)
+	if isUpdate && err != nil {
+		return &reviewResponse
+	}
+
+	err = utils.ValidateSnapshot(snapshot)
+	if err != nil {
+		reviewResponse.Allowed = false
+		reviewResponse.Result.Message = err.Error()
+		return &reviewResponse
+	}
+
+	err = checkSnapshotImmutableFields(snapshot, oldSnapshot)
+	if isUpdate && err != nil {
+		reviewResponse.Allowed = false
+		reviewResponse.Result.Message = err.Error()
+		return &reviewResponse
+	}
+
+	return &reviewResponse
+}
+
+func decideSnapshotContent(snapcontent, oldSnapcontent *volumesnapshotv1beta1.VolumeSnapshotContent, isUpdate bool) *v1.AdmissionResponse {
+	reviewResponse := v1.AdmissionResponse{
+		Allowed: true,
+		Result:  &metav1.Status{},
+	}
+
+	// if it is an UPDATE and oldSnapshot is not valid, then don't enforce strict validation
+	// This allows no-op updates to occur on snapshot resources which fail strict validation
+	// Which allows the remover of finalizers and therefore deletion of this object
+	// Don't rely on the pointers to be nil, because the deserialization method will convert it to
+	// The empty struct value. Instead check the operation type.
+	err := utils.ValidateSnapshotContent(oldSnapcontent)
+	if isUpdate && err != nil {
+		return &reviewResponse
+	}
+
+	err = utils.ValidateSnapshotContent(snapcontent)
+	if err != nil {
+		reviewResponse.Allowed = false
+		reviewResponse.Result.Message = err.Error()
+		return &reviewResponse
+	}
+
+	err = checkSnapshotContentImmutableFields(snapcontent, oldSnapcontent)
+	if isUpdate && err != nil {
+		reviewResponse.Allowed = false
+		reviewResponse.Result.Message = err.Error()
+		return &reviewResponse
+	}
+
+	return &reviewResponse
+}
+
+func stringPtrEq(s1, s2 *string) bool {
```

Comment:
might be less efficient but a cleaner implementation could be
return reflect.DeepEqual(s1, s2)
see: https://golang.org/pkg/reflect/